hathor/p2p/sync_checkpoints.py

```diff
@@ -0,0 +1,213 @@
+        prev_cp = next(it_cps)
+        for cp in it_cps:
+            self.previous_checkpoints[cp] = prev_cp
+            prev_cp = cp
+
+        # Get all checkpoints to sync
+        self.checkpoints_to_sync = [checkpoint for checkpoint in checkpoints if checkpoint.height > meta.height]
+
+        if not self.checkpoints_to_sync:
+            # Something went wrong
+            # Should start this only if there are missing checkpoints on storage
+            return
+
+        self.is_running = True
+        self.manager.reactor.callLater(0, self.try_to_sync)
+
+    # XXX: the tuple (start_hash, start_int, end_hash, end_int) appears too often, maybe make a named tuple for it
+    def _checkpoint_sync_interval(self, checkpoint: Checkpoint) -> Optional[Tuple[bytes, int, bytes, int]]:
+        """Calculate start and end point of a checkpoint."""
+        start_height, start_hash = checkpoint
+        end_height, end_hash = self.previous_checkpoints[checkpoint]
+        # XXX: this could be optimized a lot, but it actually isn't that slow
+        while start_height > end_height:
+            try:
+                block = self.manager.tx_storage.get_transaction(start_hash)
+            except TransactionDoesNotExist:
+                break
+            assert isinstance(block, Block)
+            start_hash = block.get_block_parent_hash()  # parent hash
+            start_height = block.get_metadata().height - 1  # parent height
+        # don't try to sync checkpoints that we already have all the blocks for
+        if start_height == end_height:
+            return None
+        return (start_hash, start_height, end_hash, end_height)
+
+    def _get_next_sync_interval(self) -> Optional[Tuple[bytes, int, bytes, int]]:
+        """Iterate over checkpoints_to_sync and find a valid interval to sync, pruning already synced intervals.
+
+        Will only return None when there are no more intervals to sync.
+        """
+        # for checkpoint in self.checkpoints_to_sync[::]:
+        for checkpoint in self.checkpoints_to_sync[::-1]:
+            sync_interval = self._checkpoint_sync_interval(checkpoint)
+            if sync_interval is not None:
+                return sync_interval
+            else:
+                self.checkpoints_to_sync.remove(checkpoint)
+        return None
+
+    def stop(self):
+        self.log.debug('stop sync')
+        self.is_running = False
+
+    def try_to_sync(self):
+        self.log.debug('try to sync')
+
+        if self.peer_syncing:
+            self.log.debug('already syncing')
+            return
+
+        if not self.is_running:
+            self.log.debug('not running')
+            return
+
+        if self.peers_to_request:
+            self.peer_syncing = self.peers_to_request[0]
+            self.run_sync()
+        else:
+            self.log.debug('no peers to sync from, try again later')
+            self.manager.reactor.callLater(5, self.try_to_sync)
+
+    def continue_sync(self) -> None:
```

Comment:
Why do we need this method? I feel we can let `try_sync()` handle all cases.

Reply:
What do you mean? We do let `try_sync` handle all the cases, `continue_sync` only schedules `try_sync` instead of having to do that directly where it is called.

Reply:
What do you mean? We do let `try_sync` handle all the cases, `continue_sync` only schedules `try_sync` instead of having to do that directly where it is called.

Reply:
This could be reviewed, continue_sync basically skips the current node.

Reply:
Moved to another PR.